### DISCUSSION TITLE
How to keep ib's client long-running?

### DISCUSSION THREAD

**Discussion by wihoho** [brndnmtthws/thetagang#433] [Q&A] [answered]
Hi,

Thanks for this project. One issue I have is that the ib's desk client would be closing on its own from time to time. In that case, the crontab's docker command won't be useful as I will have to start ib's client before running this binary.

Any workaround?

**Comment by brndnmtthws** [comment]
Unfortunately IBKR has made it very difficult to keep the client open, or to automate the login process. I believe the main reason they do this is that they don't want people exfiltrating the data, because the data is considered "intellectual property", i.e., an opportunity for them to collect your money.

This isn't unique to IBKR, pretty much all the brokers, exchanges, data brokers, and other 3rd party API tools use the same annoying tricks to prevent people from accessing and using the data.

This is likely the main reason they made their shitty second-factor auth so annoying, and now force everyone to use it.

It doesn't have to be this way, but it is because everyone is trying to squeeze everyone else for a few extra pennies. Perhaps one day someone with deep pockets will come along, buy the data wholesale (or buy an exchange perhaps), and liberate the data so we can have markets that aren't unfree and rigged in favour of the incumbents. More likely, however, is that we'll see bigger regulatory moats, more anti-competitive behaviour, and stricter and more hostile actions from the brokers.

So enjoy it while you still can, it will only continue to [enshittify](https://en.wiktionary.org/wiki/enshittification).

**Comment by noamagiv** [comment]
Hi, in add restart option so tws will automatically login. Manual authentication required once a week 
See https://www.ibkrguides.com/traderworkstation/auto-restart-considerations.htm

Cheers

**Comment by brndnmtthws** [comment]
> Hi, in add restart option so tws will automatically login. Manual authentication required once a week See https://www.ibkrguides.com/traderworkstation/auto-restart-considerations.htm
> 
> Cheers

FWIW, you can also run thetagang without starting the gateway (using [the `--without-ibc` CLI flag](https://github.com/brndnmtthws/thetagang/blob/f90cd7836d604f0ff95f10cf6083e08834a088ca/thetagang/main.py#L25-L30)) and have it simply connect to an existing TWS/gateway instance, if you prefer that. You just need to enable the API and use the correct host/port.

**Comment by brndnmtthws** [comment, marked as answer]
It turns out there's a way to workaround the 2FA issue for anyone else dealing with it.

You can create a second user account, provide the minimal trading permissions (i.e., disable withdrawals/transfers/management), enable the required data sources, and even set an IP restriction to limit access with those credentials. Just be sure to not enable 2FA when logging in with the second account. It's a bit annoying, but this way you can login with just a user/pass, and can reduce the damage surface area in case you lose your credentials (but just don't do that).
